Merge global_ocean_ebm into global_oce_latlon. Step1: make experiments as similar as possible

mjlosch opened the pull request:
## What changes does this PR introduce?
The main end goal of this PR is to eventually remove the verification experiment `global_ocean_ebm` (because of poor naming and some degree of redundancy) by moving and merging tests. As a first step, we merge `global_ocean_ebm` into `global_oce_latlon` to reduce clutter and misleading naming. The forward tests of `global_ocean_ebm` are copied to verification experiment `global_oce_latlon`, where the order between primary and secondary tests is changed; and the TAF-AD test of `global_ocean_ebm` will be merged into existing `global_oce_latlon` AD experiment, as a secondary "ebm" test.

## What is the current behaviour? 
- the secondary experiments in `global_ocean_ebm` do not use `pkg/ebm`, but `pkg/exf`, see also #907 
- verification experiments `global_ocean_ebm` and `global_oce_latlon` are very similar

## What is the new behaviour 
- modify both experimental setups so that they are nearly the same and can easily be merged. 
- after this, the main change is to use default `CD_CODE_OPTIONS.h`, which changes results
- cp experiments of `global_ocean_ebm` to `global_oce_latlon`, for now the directories (and reference output files) are only links (except for global_oce_latlon/input_ad.ebm, where the reference is new):
```
global_oce_latlon/input -> ../global_ocean_ebm/input.w_exf
global_oce_latlon/input.ebm -> ../global_ocean_ebm/input
global_oce_latlon/input.yearly -> ../global_ocean_ebm/input.yearly
global_oce_latlon/input_ad.ebm -> ../global_ocean_ebm/input_ad
```
Note: This merging also reconciles where former (i.e., before PR #830 and #885) experiment `global_with_exf` tests are moved to/merged into: now both forward and TAF AD tests are back in the same experiment, `global_oce_latlon`, instead of split over 2 different experiments.

## Does this PR introduce a breaking change? 
No

## Other information:
To do in a sub-sequential PR:
- move forcing files from `global_ocean_ebm/input` to  for `global_oce_latlon/input_ad.ebm`
- adjust `SIZE.h` to 6 tiles
- add TLM tests

## Suggested addition to `tag-index`
o reorganize verification experiments: 
  - adjust global_ocean_ebm-experiments and move them to global_oce_latlon
  - disable test of global_ocean_ebm (to be removed later)

mjlosch committed:
make code_ad very similar between global_oce_latlon/ocean_ebm

These changes do not change the results of the verification
experiments

mjlosch committed:
Merge upstream/master into global_oce_merge

mjlosch committed:
Merge 'upstream/master' into global_oce_merge

mjlosch committed:
add input_ad.ebm, uses global_ocean_ebm/input forcing files

matches global_ocean_ebm/results/output_adm.ebm.txt as reference if
CD_CODE_NO_AB_MOMENTUM and CD_CODE_NO_AB_CORIOLIS are defined in
CD_CODE_OPTIONS.h

mjlosch committed:
add adjMonitorFreq and adjust monitorFreq

mjlosch committed:
add reference results for AD-ebm experiment

mjlosch committed:
copy FWD experiments of global_ocean_ebm to global_oce_latlon

- global_ocean_ebm/input.w_exf  -> global_oce_latlon/input
- global_ocean_ebm/input        -> global_oce_latlon/input.ebm
- global_ocean_ebm/input.yearly -> global_oce_latlon/input.yearly

small issue: as there is already a prepare_run-script in
global_oce_latlon/input, this is also used in tr_run.ebm, I don't know
how to suppress that.

jm-c committed:
Merge branch 'master' into global_oce_merge

mjlosch commented:
1. I modified the description
2. I agree, please take care of this
3. in order to merge the AD experiments, more than one cost function needs to be defined by CPP-flags. `global_oce_latlon/code_ad/COST_OPTIONS.h` had `undef ALLOW_COST_ATLANTIC_HEAT`, which needed to be changed to `define`, so I had to turn off the cost function in `data.cost` (by `mult_atl=0.`, before it was 1., but with no effect because the code was not compiled). By the same token, I changed `mult_tracer` from 1 to 0, but that should not have any effect because the corresponding flag is not and was not defined. I didn't see any additional differences on my Mac (still comparing to reference created on villain on Sep7, 2023), probably also because there is still this issue: #910. In fact, on my Mac, the results have changed from (master)
```
Y Y Y Y 16>14<16 13 14  0  2 13 16  0  0  0  4  0  0  0 16  0  0 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
Y Y Y Y 16>15< 6 12 12  1  3 13 13  0  3  0 16  0  0  0 16  0  0 pass  global_oce_latlon.ggl90
Y Y Y Y 15>14< 7 11 11  0  2 11 11  0  1  0 13  0  0  0 13  0  0 pass  global_oce_latlon.w_exf
```
to (this PR, I removed the ebm results to avoid confusion)
```
Y Y Y Y 16>14<16 13 14 16 16 13 16 16 16 16 14 16 16 16 16 16 16 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
[...]
Y Y Y Y 16>15< 6 12 12 14 12 13 13 12 12 16 16 16 16 16 16 16 14 pass  global_oce_latlon.ggl90
Y Y Y Y 15>14< 7 11 11 13 11 11 11 11 11 16 13 12 13 16 13 13 13 pass  global_oce_latlon.w_exf
```
So I decided to go with the improvement without too many questions (o: 
On my Linux box, I don't see any differences between master and this branch for `global_oce_latlon`, either:
Master:
```
Y Y Y Y 16>16<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
Y Y Y Y 16>16<16 16 16 14 12 16 16 12 12 16 16 16 16 16 16 16 16 pass  global_oce_latlon.ggl90
Y Y Y Y 16>16<16 12 12 14 12 12 12 13 12 16 16 14 16 16 16 13 16 pass  global_oce_latlon.w_exf
```
this PR:
```
Y Y Y Y 16>16<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
[...]
Y Y Y Y 16>16<16 16 16 14 12 16 16 12 12 16 16 16 16 16 16 16 16 pass  global_oce_latlon.ggl90
Y Y Y Y 16>16<16 12 12 14 12 12 12 13 12 16 16 14 16 16 16 13 16 pass  global_oce_latlon.w_exf
```
How big are your difference?

There are differences between `global_ocean_ebm/results/output_adm.txt` and `global_oce_latlon/results/output_adm.ebm.txt` because of the different `CD_CODE_OPTIONS.h`.

jm-c committed:
remove all new input & input.* files

duplicated copy from global_ocean_ebm

jm-c committed:
add symbolic-links for new input dirs

jm-c committed:
more robust

so that it will continue to work when "global_ocean_ebm" experiment is removed

mjlosch commented:
@jm-c 
Thanks for look at this. I agree, please take care of duplicate copies in the two `input_ad` directories

The objective of all change was to make the input and code directories as similar as possible without changing the results. This includes a lot of unnecessary change and code if the experiments are kept apart, I agree. No problem to revert that again, once we have working version of the `global_ocean_ebm` experiments in `global_oce_latlon`.

The extra overlap was there in the very beginning to make `global_oce_latlon/input_ad/SIZE.h` and `global_ocean_ebm/input_ad/SIZE.h` identical. I probably fixed my MacOS problem accidentally by doing this. Before reverting to `Olx/y=2`, can we please first get the extra adjoint monitor output to see the effect of changing the overlap?

jm-c committed:
turn on monitor and AD monitor

jm-c committed:
remove all new input_ad.ebm files

jm-c committed:
add symbolic-links for new input_ad.ebm dir

jm-c committed:
back to original SIZE.h

and gerenate new output_adm.ebm.txt on ref machine

jm-c committed:
use default CD_CODE_OPTIONS.h

also do not compile exf, ggl90 and mnc pkgs
and gernerate new reference output_adm.txt (from ref. machine)

mjlosch commented:
For the record, here's what I get on MacOS:
before your changes (with my `global_oce_latlon/results/output_adm.ebm.txt`):
```
Y Y Y Y 16>14<16 13 14 16 16 13 16 16 16 16 14 16 16 16 16 16 16 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
Y Y Y Y 16>16<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 pass  global_oce_latlon.ebm
Y Y Y Y 16>15< 6 12 12 14 12 13 13 12 12 16 16 16 16 16 16 16 14 pass  global_oce_latlon.ggl90
Y Y Y Y 15>14< 7 11 11 13 11 11 11 11 11 16 13 12 13 16 13 13 13 pass  global_oce_latlon.w_exf
Y Y Y Y 15>14< 7 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- pass  global_ocean_ebm  (e=0, w=0, lfd=1, dop=1, sm=0)
```
after your changes:
```
Y Y Y Y 16>14<16 13 14  0  2 13 16  0  0  0  4  0  0  0 16  0  0 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
Y Y Y Y 16>14< 6 16 16  0  3 14 16  0  3 16 16  0  1 16 16  0  2 pass  global_oce_latlon.ebm
Y Y Y Y 16>15< 6 12 12  1  3 13 13  0  3  0 16  0  0  0 16  0  0 pass  global_oce_latlon.ggl90
Y Y Y Y 15>14< 7 11 11  0  2 11 11  0  1  0 13  0  0  0 13  0  0 pass  global_oce_latlon.w_exf
Y Y Y Y 16>14< 6 16 16 14 16 14 16 16 16 16 16 14 16 16 16 16 16 pass  global_ocean_ebm  (e=0, w=0, lfd=1, dop=1, sm=0)
```
This was kind of expected, because of issue #910, which is fixed by increasing the overlap, or by using ` -MPI 6 -command 'mpirun -n TR_NPROC ./mitgcmuv_ad'`, see here:
```
Y Y Y Y 16>14<16 13 14 16 16 13 16 16 16 16 14 16 16 16 16 16 16 pass  global_oce_latlon  (e=0, w=0, lfd=1, dop=1, sm=0)
Y Y Y Y 16>14< 6 16 16 14 16 14 16 16 16 16 16 14 16 14 16 16 16 pass  global_oce_latlon.ebm
Y Y Y Y 16>15< 6 12 12 14 12 13 13 12 12 16 16 16 16 16 16 16 14 pass  global_oce_latlon.ggl90
Y Y Y Y 15>14< 7 11 11 13 11 11 11 11 11 16 13 12 13 16 13 13 13 pass  global_oce_latlon.w_exf
```

jm-c commented:
@mjlosch I suggest to leave this MacOS sensitivity to Olx,y on the side for now, then finish the remaining, and re-consider afterward if we want to change it to 3 or keep it to 2 (or change it in an other PR). The main reason would be that if we change from a 4 tiles to a 6 tiles, it might affect this sensitivity.

And regarding the changes in `global_ocean_ebm/code_ad/`  that affect the output: You already identify what matters but I would like to list again the changes: The main one is the CD-scheme AB option (large differences); changing from 1 tile to 4 tile affects only a little bit the AD-gradient (14 digits) and the forward gradient (6 digits); and finally, having ALLOW_AUTODIFF_MONITOR (and AUTODIFF_EXCLUDE_ADEXCH_RS) defined also affect a little bit the AD-gradient (14 digits). The other changes don't seem to affect the output, but some differences in tapes (when changing CTRL_OPTIONS.h) are less obvious to check.
So I am going to bring back the original version of these files, since there is no effect on the results.

jm-c committed:
back to original version

since updated versions (from global_oce_latlon/code_ad) do not affect the output

jm-c committed:
replace fwd output with sym-link

jm-c commented:
I removed the duplicated reference FWD output (in `global_oce_latlon/results/`) and replace them with sym-links. I think we are getting closer to a "mergeable` state. @mjlosch what is your feeling ? time to remove the "work in progress" label ?

Regarding the AD tests, we still have 2 output_adm, one on each side, and will need to make a decision there. Now trying to switch the `code_ad/SIZE.h` to a 6 tiles (instead of a 4 tiles) trigger significant changes in grdchk location. So may be should be left aside, and do this in an other PR ? Or is there a chance that a 6 tiles with Olx,y =2 would fix the MacOS sensitivity issue ?

jm-c reviewed (approved):
This looks good. And suggest to postpone the upgrade of `code_ad/SIZE.h` from 4 tiles to 6 tiles to a following PR (it does require to change few `data.grdchk`) and could consider adding a Tang-Lin ref. output at the same time.

Might be useful to update PR description a little bit.

jm-c committed:
replace 1 output_adm with sym-link

+ adjust previous symbolic-links

jm-c committed:
Merge branch 'master' into global_oce_merge

jm-c committed:
disable "global_ocean_ebm" tests

and document merging of exp. "global_ocean_ebm" into "global_oce_latlon"
